### PR TITLE
Issue 96: Startup and healthcheck improvements

### LINKED ISSUE #96: Allow timing and sequencing for pod deployment for better pod readiness
Currently in Pravega-Operator dependent pods (like Segementstore depends on other pods / svcs Bookie and Zookeeper) sometime get initialization failures during initial deployment if the dependent pods are not yet ready / getting ready. 

In a sample deployment like below with following pravega.yaml

```
   apiVersion: "pravega.pravega.io/v1alpha1"
   kind: "PravegaCluster"
   metadata:
     name: "pravega"
   spec:
     zookeeperUri: [ZK IP]:2181`
```
```
   bookkeeper:
    image:
      repository: deenadock/bookkeeper
      tag: 0.4.0-rc1
      pullPolicy: IfNotPresent
```

    replicas: 4

    storage:
      ledgerVolumeClaimTemplate:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "standard"
        resources:
          requests:
            storage: 20Gi

      journalVolumeClaimTemplate:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "standard"
        resources:
          requests:
            storage: 20Gi

    autoRecovery: true

```
  pravega:
    controllerReplicas: 2
    segmentStoreReplicas: 3
```

    cacheVolumeClaimTemplate:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: "standard"
      resources:
        requests:
          storage: 30Gi

    image:
      repository: deenadock/pravega
      tag: 0.4.0-rc1
      pullPolicy: IfNotPresent

    tier2:
      filesystem:
        persistentVolumeClaim:
          claimName: pravega-tier2

During this deployment seems like Bookie and Segmentstore is getting into _ContainerCreating_ state at the same time and as Segementstore have dependency on Bookie to be up it is failing and getting restarted couple of times with following exception.

```
kubectl get po --all-namespaces
NAMESPACE     NAME                                                        READY     STATUS              RESTARTS   AGE
default       interested-pike-nfs-server-provisioner-0                    1/1       Running             0          8m
default       pravega-bookie-0                                            0/1       ContainerCreating   0          34s
default       pravega-bookie-1                                            0/1       ContainerCreating   0          33s
default       pravega-bookie-2                                            0/1       ContainerCreating   0          33s
default       pravega-bookie-3                                            0/1       ContainerCreating   0          33s
default       pravega-example-zk-0                                        1/1       Running             0          19m
default       pravega-example-zk-1                                        1/1       Running             0          18m
default       pravega-example-zk-2                                        1/1       Running             0          17m
default       pravega-operator-785cf444dd-t2q69                           1/1       Running             0          11m
default       pravega-pravega-controller-7c7d5d45cd-fdbfx                 1/1       Running             0          33s
default       pravega-pravega-controller-7c7d5d45cd-rpw9s                 1/1       Running             0          33s
default       pravega-segmentstore-0                                      0/1       Error               1          33s
default       pravega-segmentstore-1                                      0/1       ContainerCreating   0          33s
default       pravega-segmentstore-2                                      0/1       ContainerCreating   0          33s
```

 ```
09:06:31,724 1001 [main] ERROR i.p.s.server.host.ServiceStarter - Could not start the Service, Aborting.
io.pravega.segmentstore.storage.DataLogNotAvailableException: Unable to establish connection to ZooKeeper or BookKeeper.
        at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogFactory.initialize(BookKeeperLogFactory.java:96)
        at io.pravega.segmentstore.server.store.ServiceBuilder.initialize(ServiceBuilder.java:230)
        at io.pravega.segmentstore.server.host.ServiceStarter.start(ServiceStarter.java:96)
        at io.pravega.segmentstore.server.host.ServiceStarter.main(ServiceStarter.java:228)
Caused by: org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /pravega/pravega/bookkeeper/ledgers/available/readonly
        at org.apache.zookeeper.KeeperException.create(KeeperException.java:115)
        at org.apache.zookeeper.KeeperException.create(KeeperException.java:51)
        at org.apache.zookeeper.ZooKeeper.create(ZooKeeper.java:1426)
        at org.apache.bookkeeper.zookeeper.ZooKeeperClient.access$1801(ZooKeeperClient.java:71)
        at org.apache.bookkeeper.zookeeper.ZooKeeperClient$9.call(ZooKeeperClient.java:695)
        at org.apache.bookkeeper.zookeeper.ZooKeeperClient$9.call(ZooKeeperClient.java:689)
        at org.apache.bookkeeper.zookeeper.ZooWorker.syncCallWithRetries(ZooWorker.java:141)
        at org.apache.bookkeeper.zookeeper.ZooKeeperClient.create(ZooKeeperClient.java:689)
        at org.apache.bookkeeper.client.BookieWatcher$ReadOnlyBookieWatcher.<init>(BookieWatcher.java:345)
        at org.apache.bookkeeper.client.BookieWatcher.<init>(BookieWatcher.java:98)
        at org.apache.bookkeeper.client.BookKeeper.<init>(BookKeeper.java:486)
        at org.apache.bookkeeper.client.BookKeeper.<init>(BookKeeper.java:321)
        at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogFactory.startBookKeeperClient(BookKeeperLogFactory.java:156)
        at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogFactory.initialize(BookKeeperLogFactory.java:84)
        ... 3 common frames omitted

```
It would be good to have a timing and sequencing involved in pod deployment so that dependent pods would be deployed after necessary parent pods are running properly.

### PR DESCRIPTION
**Pull request by adrianmo**
#### Readiness probes
[Readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-readiness-probes) have been added to prevent the pods from becoming _ready_ until they have started up correctly and are actually accepting requests. 

Bookkeeper and Controller pods should start quickly because they only depend on ZooKeeper, which should be already up and ready. The maximum time allowed for them to become ready is 1.5 and 1 minute respectively.

Segment Store pods depend on BookKeeper, and if configured with external access enabled, they will have to wait for the service external IP to be allocated. In GKE, we have experienced up to three minutes for the external IP to be allocated and obtained by the pod, therefore, the timeout to become ready has been set to 5 minutes. 

The command to verify pod readiness is the same for all components and checks whether or not there's a process listening on the relevant port.

#### Liveness probe
In addition to the Readiness probes, [Liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-command) have been added to detect and remedy situations in which the pod transitions to a broken state. If the pod becomes unhealthy, the liveness probe will detect it and restart it. The command to verify pod liveness is the same used for readiness.

#### Segment Store wait for Bookkeeper on startup
Starting Segment Store and BookKeeper pods at the same time was causing Segment Store instances to throw errors and restart a couple of times until Bookies were up and accepting requests. 

Pravega Docker images accept a [`WAIT_FOR` environment variable](https://github.com/pravega/pravega/blob/master/docker/compose/docker-compose.yml#L89) which tells the container to wait for one or more systems to become reachable until the container is allowed to start. This PR leverages that option to configure Segment Store pods to wait until at least three Bookies become ready.

#### Segment Store sequential, ordered startup
To avoid errors in the bootstrap process, Segment Store pods will now start sequentially, one after another. This is also aligned with how external IP are sequentially allocated and with the new readiness and liveness probes, i.e., if we were to start all Segment Store in parallel, at the same time, the pods with higher rank would time out their readiness probes before their external IP addresses are allocated. 

## Purpose of the change
Fix #96 

## How to test the change
Checkout the repository (`master` branch is fine, no need to checkout this branch).

Edit the operator image used in the `deploy/operator.yaml` file to use the `adrianmo/pravega-operator:issue-96-6` image, which contains the changes in this PR.

Deploy the operator.

```
kubectl create -f deploy/operator.yaml
```

Follow the instructions in the README file to deploy ZooKeeper, NFS provisioner, and a Pravega cluster with 3 Bookies, 1 or more Controllers, and 3 Segment stores.

When you deploy a Pravega cluster, check that Segment Store instances are created sequentially.

```
$ kubectl get po -l pravega_cluster=pravega
NAME                                          READY     STATUS    RESTARTS   AGE
pravega-bookie-0                              1/1       Running   0          44s
pravega-bookie-1                              1/1       Running   0          44s
pravega-bookie-2                              1/1       Running   0          44s
pravega-pravega-controller-5c7556c797-x5694   1/1       Running   0          44s
pravega-pravega-segmentstore-0                0/1       Running   0          44s
```

Check the logs of Segment Store 0 pod to see how it tries to connect to the Bookies.

```
$ kubectl logs pravega-pravega-segmentstore-0
Waiting for pravega-bookie-0.pravega-bookie-headless.default:3181
Unable to connect to pravega-bookie-0.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Unable to connect to pravega-bookie-0.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Unable to connect to pravega-bookie-0.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Connected to pravega-bookie-0.pravega-bookie-headless.default:3181
Waiting for pravega-bookie-1.pravega-bookie-headless.default:3181
Unable to connect to pravega-bookie-1.pravega-bookie-headless.default:3181. Waiting
[Errno -2] Name does not resolve
Connected to pravega-bookie-1.pravega-bookie-headless.default:3181
Waiting for pravega-bookie-2.pravega-bookie-headless.default:3181
Connected to pravega-bookie-2.pravega-bookie-headless.default:3181
...
```

As soon as `pravega-pravega-segmentstore-0` becomes _ready_, `pravega-pravega-segmentstore-1` should be created.

```
$ kubectl get po -l pravega_cluster=pravega
NAME                                          READY     STATUS    RESTARTS   AGE
pravega-bookie-0                              1/1       Running   0          3m
pravega-bookie-1                              1/1       Running   0          3m
pravega-bookie-2                              1/1       Running   0          3m
pravega-pravega-controller-5c7556c797-x5694   1/1       Running   0          3m
pravega-pravega-segmentstore-0                1/1       Running   0          3m
pravega-pravega-segmentstore-1                0/1       Running   0          20s
```

Inspect all pods to check that Liveness and Readiness have been set successfully.

```
$ kubectl describe pod pravega-bookie-1 |grep 'Liveness\|Readiness'
    Liveness:   exec [/bin/sh -c netstat -ltn | grep 3181 || ss -ltn | grep 3181] delay=60s timeout=1s period=15s #success=1 #failure=4
    Readiness:  exec [/bin/sh -c netstat -ltn | grep 3181 || ss -ltn | grep 3181] delay=0s timeout=1s period=10s #success=1 #failure=9
```

### CODE REVIEW

**Comment by adrianmo** [comment]
[Operator: Configure readiness and liveness probes (healthchecks)](https://trello.com/c/5WmZ3k5O/41-operator-configure-readiness-and-liveness-probes-healthchecks)

**Comment by adrianmo** [comment]
Need to rebase and fix conflicts.

**Comment by adrianmo** [comment]
PR is now up to date with master. Ready to review.

File: pkg/controller/pravega/pravega_segmentstore.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ func MakeSegmentStoreStatefulSet(pravegaCluster *api.PravegaCluster) *appsv1.Sta
 		Spec: appsv1.StatefulSetSpec{
 			ServiceName:         "pravega-segmentstore",
 			Replicas:            &pravegaCluster.Spec.Pravega.SegmentStoreReplicas,
-			PodManagementPolicy: appsv1.ParallelPodManagement,
+			PodManagementPolicy: appsv1.OrderedReadyPodManagement,
```
Comment (fpj) [inline pkg/controller/pravega/pravega_segmentstore.go]: Why have you changed from parallel to ordered?

File: pkg/util/pravegacluster.go [hunk 2]
```diff
@@ -79,3 +79,15 @@ func LabelsForPravegaCluster(pravegaCluster *v1alpha1.PravegaCluster, component
 func PravegaControllerServiceURL(pravegaCluster v1alpha1.PravegaCluster) string {
 	return fmt.Sprintf("tcp://%v.%v:%v", ServiceNameForController(pravegaCluster.Name), pravegaCluster.Namespace, "9090")
 }
+
+func HealthcheckCommand(port int32) []string {
+	return []string{"/bin/sh", "-c", fmt.Sprintf("netstat -ltn | grep %d || ss -ltn | grep %d", port, port)}
```
Comment (fpj) [inline pkg/util/pravegacluster.go]: Could you use `nc` here to check the status of the port?

**Comment by adrianmo** [comment]
Test cases will be added in #92.

File: deploy/operator.yaml [hunk 3]
```diff
@@ -15,7 +15,7 @@ spec:
       serviceAccountName: pravega-operator
       containers:
         - name: pravega-operator
-          image: pravega/pravega-operator:latest
+          image: adrianmo/pravega-operator:issue-96
```
Comment (fpj) [inline deploy/operator.yaml]: Is it right that you want this to point to this issue/snapshot image?

File: test/e2e/basic_test.go [hunk 4]
```diff
@@ -14,9 +14,33 @@ import (
...
 
 	framework "github.com/operator-framework/operator-sdk/pkg/test"
+
+	pravega_e2eutil "github.com/pravega/pravega-operator/pkg/test/e2e/e2eutil"
 )
 
-func testCreateCluster(t *testing.T, f *framework.Framework, ctx *framework.TestCtx) error {
-	t.Skip("Not implemented")
-	return nil
+func testCreateDefaultCluster(t *testing.T) {
```
Comment (fpj) [inline test/e2e/basic_test.go]: This test seems to be end-to-end. I'm wondering if we should be doing more focused test cases, along the lines of unit tests.

File: deploy/operator.yaml (hunk 3 above)
Comment (adrianmo) [reply deploy/operator.yaml]: Thanks for raising this, I've reverted the image to `pravega/pravega-operator:latest`.

File: pkg/test/e2e/e2eutil/pravegacluster_util.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+// DeleteCluster deletes the PravegaCluster CR specified by cluster spec
+func DeleteCluster(t *testing.T, f *framework.Framework, ctx *framework.TestCtx, p *api.PravegaCluster) error {
+	t.Logf("deleting pravega cluster: %v", p.Name)
+	err := f.Client.Delete(goctx.TODO(), p)
+	if err != nil {
+		return fmt.Errorf("failed to delete CR: %v", err)
+	}
+	// TODO: Wait for cluster resources to be deleted
```
[the remark this answers is not in the document]
Comment (adrianmo) [reply pkg/test/e2e/e2eutil/pravegacluster_util.go]: I've added a function to wait for the cluster to terminate.

File: test/e2e/basic_test.go (hunk 4 above)
Comment (adrianmo) [reply test/e2e/basic_test.go]: @fpj currently there's not a single unit test in the project. I've created issue https://github.com/pravega/pravega-operator/issues/113 to start adding coverage to the project.

**Comment by adrianmo** [comment]
@fpj @sumit-bm this PR is ready for second round of reviews after adding E2E test cases.

**Pull request merged**